devlib/utils/android.py

```diff
@@ -508,3 +509,89 @@ def _check_env():
         platform_tools = _env.platform_tools
         adb = _env.adb
         aapt = _env.aapt
+
+class LogcatMonitor(threading.Thread):
+    def __init__(self, device, regexps=['.*']):
+        super(LogcatMonitor, self).__init__()
+
+        self.device = device
+
+        self._stopped = threading.Event()
+        self._seeking = threading.Event()
+
+        self._seeked = None
+        self._found = None
+
+        self._lines = []
+        self._datalock = threading.Lock()
+        self._regexps = regexps
+
+    def run(self):
+        self.device.clear_logcat()
+
+        # Join all requested regexps with an 'or'
+        regexp = '{}'.format('|'.join(self._regexps))
+        if len(self._regexps) > 1:
+            regexp = '({})'.format(regexp)
+
+        self._logcat = self.device.background('logcat -e {}'.format(regexp))
+
+        while not self._stopped.is_set():
+            line = self._logcat.stdout.readline(1024)
+            self._add_line(line)
+
+    def _add_line(self, line):
+        self._datalock.acquire()
```

Comment:
Rather than doing explicit acquire/release, you can use the lock as a context:

```
with self._datalock:
       # do stuff
```

This is cleaner and will also ensure that the lock is released should an exception occur with it held.